src/site/content/en/blog/porting-libusb-to-webusb/index.md

```diff
@@ -0,0 +1,450 @@
+
+```cpp
+namespace {
+// clang-format off
+  EM_JS(EM_VAL, em_promise_catch_impl, (EM_VAL handle), {
+    let promise = Emval.toValue(handle);
+    promise = promise.then(
+      value => ({error : 0, value}),
+      error => {
+        const ERROR_CODES = {
+          // LIBUSB_ERROR_IO
+          NetworkError : -1,
+          // LIBUSB_ERROR_INVALID_PARAM
+          DataError : -2,
+          TypeMismatchError : -2,
+          IndexSizeError : -2,
+          // LIBUSB_ERROR_ACCESS
+          SecurityError : -3,
+          …
+        };
+        console.error(error);
+        let errorCode = -99; // LIBUSB_ERROR_OTHER
+        if (error instanceof DOMException)
+        {
+          errorCode = ERROR_CODES[error.name] ?? errorCode;
+        }
+        else if (error instanceof RangeError || error instanceof TypeError)
+        {
+          errorCode = -2; // LIBUSB_ERROR_INVALID_PARAM
+        }
+        return {error: errorCode, value: undefined};
+      }
+    );
+    return Emval.toHandle(promise);
+  });
+// clang-format on
+
+val em_promise_catch(val &&promise) {
+  EM_VAL handle = promise.as_handle();
+  handle = em_promise_catch_impl(handle);
+  return val::take_ownership(handle);
+}
+
+// C++ struct representation for {value, error} object from above
+// (performs conversion in the constructor).
+struct promise_result {
+  libusb_error error;
+  val value;
+
+  promise_result(val &&result)
+      : error(static_cast<libusb_error>(result["error"].as<int>())),
+        value(result["value"]) {}
+
+  // C++ counterpart of the promise helper above that takes a promise, catches
+  // its error, converts to a libusb status and returns the whole thing as
+  // `promise_result` struct for easier handling.
+  static promise_result await(val &&promise) {
+    promise = em_promise_catch(std::move(promise));
+    return {promise.await()};
+  }
+};
+```
+
+Now I could use `promise_result::await` on any `Promise` returned from WebUSB operations and inspect its `error` and `value` fields separately.
+
+For example, retrieving a `val` representing a [`USBDevice`](https://developer.mozilla.org/docs/Web/API/USBDevice) from `libusb_device_handle`, calling its [`.open()`](https://developer.mozilla.org/docs/Web/API/USBDevice/open) method, awaiting its result and returning an error code as a libusb status code looks like this:
```

Comment:
```suggestion
For example, retrieving a `val` representing a [`USBDevice`](https://developer.mozilla.org/docs/Web/API/USBDevice) from `libusb_device_handle`, calling its [`open()`](https://developer.mozilla.org/docs/Web/API/USBDevice/open) method, awaiting its result, and returning an error code as a libusb status code looks like this:
```

Reply:
We don't include the period for standalone method references. We do use a comma before the 'and' for items in a list (Oxford comma).

Reply:
Ah interesting, I find `.` distinction between standalone function vs methods helpful, but, of course, general styleguide is more important.